File: java/src/Closure.java
@java.lang.FunctionalInterface
interface Counter {
    int inc();
}

public class Closure {
    static Counter makeCounter() {
        final int[] container = { 0 };
        return () -> (++container[0]);
    }
    public static void main(String[] args) {
        Counter counter = makeCounter();
        System.out.println(counter.inc());
        System.out.println(counter.inc());
        System.out.println(counter.inc());
    }
}

File: golang/src/closure.go
package main

import "fmt"

func makeCounter() func() int {
	var i = 0
	return func() int {
		i++
		return i
	}
}

func main() {
	counter := makeCounter()
	fmt.Println(counter()) // 1
	fmt.Println(counter()) // 2
	fmt.Println(counter()) // 3
}
